Fix traversal for moved out elements by mapping placeholders

chirags27 opened the pull request:
#### 0ff58f04d04d4c6b1aa1e2635dce38903db82f9a
<pre>
Fix traversal for moved out elements by mapping placeholders
<a href="https://bugs.webkit.org/show_bug.cgi?id=251126">https://bugs.webkit.org/show_bug.cgi?id=251126</a>
rdar://94017781

Reviewed by Alan Baradlay.

Before this change, when we used to traverse the ancestor chain for
autoinline validation in
propagateRepaintToParentWithOutlineAutoIfNeeded, we just looked at the
direct parent of the moved out element. That was incorrect, because the
moved out elements have a placeholder inserted in their original
position and hence we need to look it up in the enclosing scope. This
change fixes that.

* LayoutTests/fast/rendering/render-multi-column-set-invalidate-crash-expected.txt: Added.
* LayoutTests/fast/rendering/render-multi-column-set-invalidate-crash.html: Added.
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
(WebCore::RenderLayerModelObject::willBeRemovedFromTree):
* Source/WebCore/rendering/RenderMultiColumnFlow.cpp:
(WebCore::RenderMultiColumnFlow::findColumnSpannerPlaceholder const):
* Source/WebCore/rendering/RenderMultiColumnFlow.h:
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::propagateRepaintToParentWithOutlineAutoIfNeeded const):
* Source/WebCore/rendering/updating/RenderTreeBuilder.cpp:
(WebCore::RenderTreeBuilder::detachFromRenderElement):

Canonical link: <a href="https://commits.webkit.org/259412@main">https://commits.webkit.org/259412@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/31381c434efec626b9439aad18c372fb96b589d3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104790 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13870 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37694 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114066 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174268 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108703 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15003 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4799 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97123 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112980 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110550 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11596 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94607 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39123 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93456 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26221 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/80787 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7222 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27581 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7321 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4174 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13373 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47131 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6490 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9108 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

chirags27 commented on LayoutTests/fast/rendering/render-multi-column-set-invalidate-crash.html:
```diff
@@ -0,0 +1,74 @@
+<head>
```
This is the reduced version of the test (see the original one in the radar)

I've tried a few things to reduce it further by removing the event handlers, but the crash doesn't repro with that, so if you folks have ideas please let me know :)

alanbaradlay commented on Source/WebCore/rendering/updating/RenderTreeBuilder.cpp:
```diff
@@ -942,10 +942,11 @@ RenderPtr<RenderObject> RenderTreeBuilder::detachFromRenderElement(RenderElement
     // So that we'll get the appropriate dirty bit set (either that a normal flow child got yanked or
     // that a positioned child got yanked). We also repaint, so that the area exposed when the child
     // disappears gets repainted properly.
+    bool shouldNotRepaint = child.previousSibling() && is<RenderMultiColumnSet>(child.previousSibling());
```
I'd move this inside "if (!parent.renderTreeBeingDestroyed()" and skip both repaints.

alanbaradlay replied:
also is<> can take null_ptr

alanbaradlay commented on Source/WebCore/rendering/RenderLayerModelObject.cpp:
```diff
@@ -84,7 +85,8 @@ void RenderLayerModelObject::willBeDestroyed()
 void RenderLayerModelObject::willBeRemovedFromTree(IsInternalMove isInternalMove)
 {
-    if (auto* layer = this->layer(); layer && layer->needsFullRepaint() && isInternalMove == IsInternalMove::No)
+    bool shouldNotRepaint = this->previousSibling() && is<RenderMultiColumnSet>(this->previousSibling());
```
is<> can take null_ptr and I'd not put this in-between the layer checks, but maybe as a trailing condition

alanbaradlay commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -915,6 +915,15 @@ void RenderObject::propagateRepaintToParentWithOutlineAutoIfNeeded(const RenderL
     // Issue repaint on the renderer with outline: auto.
     for (const auto* renderer = this; renderer; renderer = renderer->parent()) {
+        if (renderer->previousSibling() && is<RenderMultiColumnSet>(renderer->previousSibling())) {
+            auto prevMultiColumnSet = downcast<RenderMultiColumnSet>(renderer->previousSibling());
```
please do not abbreviate

alanbaradlay commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -915,6 +915,15 @@ void RenderObject::propagateRepaintToParentWithOutlineAutoIfNeeded(const RenderL
     bool repaintRectNeedsConverting = false;
     // Issue repaint on the renderer with outline: auto.
     for (const auto* renderer = this; renderer; renderer = renderer->parent()) {
+        if (renderer->previousSibling() && is<RenderMultiColumnSet>(renderer->previousSibling())) {
```
should just use is<>
or
if (auto* previousMultiColumnSet = renderer->previousSibling(); is<RenderMultiColumnSet(previousMultiColumnSet))

alanbaradlay commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -915,6 +915,15 @@ void RenderObject::propagateRepaintToParentWithOutlineAutoIfNeeded(const RenderL
+        if (renderer->previousSibling() && is<RenderMultiColumnSet>(renderer->previousSibling())) {
+            auto prevMultiColumnSet = downcast<RenderMultiColumnSet>(renderer->previousSibling());
+            auto enclosingMultiColumnFlow = prevMultiColumnSet->multiColumnFlow();
+            auto& prevMultiColumnSetBox = downcast<RenderBox>(*renderer);
```
hope we don't move out non-RenderBox type of renderers (I'd test it with a <span> -RenderInline just to make sure this downcast is valid)

chirags27 replied:
Verified

nt1m commented on LayoutTests/fast/rendering/render-multi-column-set-invalidate-crash.html:
```diff
@@ -0,0 +1,74 @@
+<style>
+:not(bgsound) { -webkit-column-rule: solid; column-span: all; }
+.class1 { display: flex; float: right; black; -webkit-user-modify: read-write; }
+* { green; -webkit-column-count: 44; }
```
```suggestion
* { column-rule: solid; column-span: all; column-count: 44; }
.class1 { display: flex; float: right; -webkit-user-modify: read-write; }
```

There's no bgsound in the file, so `:not(bgsound)` is just `*`. Although I doubt we need to target all elements on the page. The invalid `black;` and `green;` can be removed.

nt1m commented on LayoutTests/fast/rendering/render-multi-column-set-invalidate-crash.html:
```diff
@@ -0,0 +1,74 @@
+    try {
+        elementID3.setAttribute("hidden", "hidden"); 
+        container = document.caretRangeFromPoint(294,50).startContainer; 
+        document.all[10%document.all.length].appendChild(elementID1);
```
I'm sure `10%document.all.length` can be replaced with a static number. You can probably find it by computing it from the Web Inspector.

It's better just to refer to the relevant element directly.

nt1m commented on LayoutTests/fast/rendering/render-multi-column-set-invalidate-crash.html:
```diff
@@ -0,0 +1,74 @@
+        testRunner.waitUntilDone();
+    }
+    try {
+        elementID3.setAttribute("hidden", "hidden"); 
```
```suggestion
        elementID3.hidden = true;
```

nt1m commented on LayoutTests/fast/rendering/render-multi-column-set-invalidate-crash.html:
```diff
@@ -0,0 +1,74 @@
+        elementID3.setAttribute("hidden", "hidden"); 
+        container = document.caretRangeFromPoint(294,50).startContainer; 
+        document.all[10%document.all.length].appendChild(elementID1);
+        container.setAttribute("onwebkitfullscreenerror", "eventhandler1()");
```
`container.addEventListener("webkitfullscreenerror", eventhandler1, { once: true });`

nt1m commented on LayoutTests/fast/rendering/render-multi-column-set-invalidate-crash.html:
```diff
@@ -0,0 +1,74 @@
+            }
+            return;
+        }
+        var markElement = document.createElement("mark"); //HTMLUnknownElement
```
```suggestion
        let unknownElement = document.createElement("unknown");
```

nt1m commented on LayoutTests/fast/rendering/render-multi-column-set-invalidate-crash.html:
```diff
@@ -0,0 +1,74 @@
+    try {
+        elementID2.prepend(String.fromCodePoint(0));
+        elementID4.length = 99;
+        document.all[91%document.all.length].appendChild(elementID6);
```
ditto here.

nt1m commented on LayoutTests/fast/rendering/render-multi-column-set-invalidate-crash.html:
```diff
@@ -0,0 +1,74 @@
+
+    try {
+        elementID2.prepend(String.fromCodePoint(0));
+        elementID4.length = 99;
```
I'd be surprised if that line did anything.

nt1m commented on LayoutTests/fast/rendering/render-multi-column-set-invalidate-crash.html:
```diff
@@ -0,0 +1,74 @@
+function eventhandler1() {
+    try {
+        runcount["eventhandler1"]++;
+        if(runcount["eventhandler1"] > 2) {
```
This is kinda funky because the first event listener is webkitfullscreenevent, then it's followed by requestAnimationFrame. I think we should remove the runcount logic, at least for eventhandler1.

I think the part where you do `runcount["eventhandler1"] > 2` can be factored out into a new function, and that function can be referred to from `requestAnimationFrame()` below

chirags27 committed:
Fix traversal for moved out elements by mapping placeholders
https://bugs.webkit.org/show_bug.cgi?id=251126
rdar://94017781

Reviewed by Alan Baradlay.

Before this change, when we used to traverse the ancestor chain for
autoinline validation in
propagateRepaintToParentWithOutlineAutoIfNeeded, we just looked at the
direct parent of the moved out element. That was incorrect, because the
moved out elements have a placeholder inserted in their original
position and hence we need to look it up in the enclosing scope. This
change fixes that.

* LayoutTests/fast/rendering/render-multi-column-set-invalidate-crash-expected.txt: Added.
* LayoutTests/fast/rendering/render-multi-column-set-invalidate-crash.html: Added.
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
(WebCore::RenderLayerModelObject::willBeRemovedFromTree):
* Source/WebCore/rendering/RenderMultiColumnFlow.cpp:
(WebCore::RenderMultiColumnFlow::findColumnSpannerPlaceholder const):
* Source/WebCore/rendering/RenderMultiColumnFlow.h:
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::propagateRepaintToParentWithOutlineAutoIfNeeded const):
* Source/WebCore/rendering/updating/RenderTreeBuilder.cpp:
(WebCore::RenderTreeBuilder::detachFromRenderElement):

Canonical link: https://commits.webkit.org/259412@main
